staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```diff
@@ -0,0 +1,216 @@
+	syncFn func(string) error
+
+	queue workqueue.RateLimitingInterface
+
+	staticSpec     *spec.Swagger
+	openAPIService *handler.OpenAPIService
+
+	// specs per version and per CRD name
+	lock     sync.Mutex
+	crdSpecs map[string]map[string]*spec.Swagger
+}
+
+// NewController creates a new Controller with input CustomResourceDefinition informer
+func NewController(crdInformer informers.CustomResourceDefinitionInformer) *Controller {
+	c := &Controller{
+		crdLister:  crdInformer.Lister(),
+		crdsSynced: crdInformer.Informer().HasSynced,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crd_openapi_controller"),
+		crdSpecs:   map[string]map[string]*spec.Swagger{},
+	}
+
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCustomResourceDefinition,
+		UpdateFunc: c.updateCustomResourceDefinition,
+		DeleteFunc: c.deleteCustomResourceDefinition,
+	})
+
+	c.syncFn = c.sync
+	return c
+}
+
+// Run sets openAPIAggregationManager and starts workers
+func (c *Controller) Run(staticSpec *spec.Swagger, openAPIService *handler.OpenAPIService, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	defer klog.Infof("Shutting down OpenAPI controller")
+
+	klog.Infof("Starting OpenAPI controller")
+
+	c.staticSpec = staticSpec
+	c.openAPIService = openAPIService
+
+	if !cache.WaitForCacheSync(stopCh, c.crdsSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	// only start one worker thread since its a slow moving API
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	// log slow aggregations
+	start := time.Now()
+	defer func() {
+		elapsed := time.Since(start)
+		if elapsed > time.Second {
+			klog.Warningf("slow openapi aggregation of %q: %s", key.(string), elapsed)
+		}
+	}()
+
+	err := c.syncFn(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+	return true
+}
+
+func (c *Controller) sync(name string) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	crd, err := c.crdLister.Get(name)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// do we have to remove all specs of this CRD?
+	if errors.IsNotFound(err) || !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) {
+		if _, found := c.crdSpecs[name]; !found {
+			return nil
+		}
+		delete(c.crdSpecs, name)
+		return c.updateSpec()
+	}
+
+	// compute CRD spec and see whether it changed
+	oldSpecs := c.crdSpecs[crd.Name]
+	newSpecs := map[string]*spec.Swagger{}
+	anyChanged := false
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
```

Comment:
if it was previously served and is no longer served, oldSpecs could contain stale versions, right? need to check if oldSpecs[v.Name] is set, and if so, delete it and set anyChanged = true.

Reply:
also, add a test that would have caught this (flip a previously served version to unserved and ensure it gets removed from oldSpecs)

Reply:
we check `len(oldSpecs) == len(newSpecs)` so it should be caught. will add a test for this

Reply:
> we check len(oldSpecs) == len(newSpecs) so it should be caught

that would catch that it needed updating, but it would never be removed